### ISSUE TITLE
[ppc64el] Build failure on ppc64el (power 8)

### ISSUE THREAD

**Issue by cdluminate** [flame/blis#258]
https://launchpadlibrarian.net/390523995/buildlog_ubuntu-cosmic-ppc64el.blis_0.4.1-1~u2_BUILDING.txt.gz

The `power7` config seems not working. Or should I use `generic` for powerpc64el architecture instead?

**Comment by fgvanzee** [comment]
@cdluminate That sounds fine. We don't have power7 hardware to test on, and that configuration has probably not been tested in years.

**Comment by jeffhammond** [comment]
`generic` doesn't work either.

# GCC
```sh
[jhammond@vestalac2 blis]$ ./configure
configure: detected Linux kernel version 2.6.32-754.3.5.el6.ppc64.
configure: C compiler search list is: gcc clang cc.
configure: using 'gcc' C compiler.
configure: C++ compiler search list is: g++ clang++ c++.
configure: using 'g++' C++ compiler (for sandbox only).
configure: found gcc version 4.4.7 (maj: 4, min: 4, rev: 7).
configure: checking for blacklisted configurations due to gcc 4.4.7.
configure: warning: gcc 4.4.7 does not support 'knl'; adding to blacklist.
configure: error: configure: *** Unsupported compiler version: gcc 4.4.7.
[jhammond@vestalac2 blis]$ ./configure generic
configure: detected Linux kernel version 2.6.32-754.3.5.el6.ppc64.
configure: C compiler search list is: gcc clang cc.
configure: using 'gcc' C compiler.
configure: C++ compiler search list is: g++ clang++ c++.
configure: using 'g++' C++ compiler (for sandbox only).
configure: found gcc version 4.4.7 (maj: 4, min: 4, rev: 7).
configure: checking for blacklisted configurations due to gcc 4.4.7.
configure: warning: gcc 4.4.7 does not support 'knl'; adding to blacklist.
configure: error: configure: *** Unsupported compiler version: gcc 4.4.7.
```

# Clang
```
[jhammond@vestalac2 blis]$ ./configure CC=clang generic
configure: detected Linux kernel version 2.6.32-754.3.5.el6.ppc64.
configure: C compiler search list is: gcc clang cc.
configure: using 'clang' C compiler.
configure: C++ compiler search list is: g++ clang++ c++.
configure: using 'g++' C++ compiler (for sandbox only).
configure: found clang version 4.2.1 (maj: 4, min: 2, rev: 1).
configure: checking for blacklisted configurations due to clang 4.2.1.
configure: found assembler ('as') version 2.20.51 (maj: 2, min: 20, rev: 51).
configure: checking for blacklisted configurations due to as 2.20.51.
configure: warning: assembler ('as' 2.20.51) does not support 'bulldozer'; adding to blacklist.
configure: warning: assembler ('as' 2.20.51) does not support 'sandybridge'; adding to blacklist.
configure: warning: assembler ('as' 2.20.51) does not support 'haswell'; adding to blacklist.
configure: warning: assembler ('as' 2.20.51) does not support 'piledriver'; adding to blacklist.
configure: warning: assembler ('as' 2.20.51) does not support 'steamroller'; adding to blacklist.
configure: warning: assembler ('as' 2.20.51) does not support 'excavator'; adding to blacklist.
configure: warning: assembler ('as' 2.20.51) does not support 'skx'; adding to blacklist.
configure: warning: assembler ('as' 2.20.51) does not support 'knl'; adding to blacklist.
configure: configuration blacklist:
configure:   bulldozer sandybridge haswell piledriver steamroller excavator skx knl
configure: reading configuration registry...done.
configure: determining default version string.
configure: found '.git' directory; assuming git clone.
configure: executing: git describe --tags.
configure: got back 0.4.1-21-g807a6548.
configure: truncating to 0.4.1-21.
configure: starting configuration of BLIS 0.4.1-21.
configure: configuring with official version string.
configure: found shared library .so version '1.0.0'.
configure:   .so major version: 1
configure:   .so minor.build version: 0.0
configure: manual configuration requested; configuring with 'generic'.
configure: checking configuration against contents of 'config_registry'.
configure: configuration 'generic' is registered.
configure: 'generic' is defined as having the following sub-configurations:
configure:    generic
configure: which collectively require the following kernels:
configure:    generic
configure: checking sub-configurations:
configure:   'generic' is registered...and exists.
configure: checking sub-configurations' requisite kernels:
configure:   'generic' kernels...exist.
configure: no install prefix option given; defaulting to '/home/jhammond/blis'.
configure: no install libdir option given; defaulting to PREFIX/lib.
configure: no install includedir option given; defaulting to PREFIX/include.
configure: no install sharedir option given; defaulting to PREFIX/share.
configure: final installation directories:
configure:   libdir:     /home/jhammond/blis/lib
configure:   includedir: /home/jhammond/blis/include
configure:   sharedir:   /home/jhammond/blis/share
configure: no preset CFLAGS detected.
configure: no preset LDFLAGS detected.
configure: debug symbols disabled.
configure: disabling verbose make output. (enable with 'make V=1'.)
configure: disabling ARG_MAX hack.
configure: building BLIS as both static and shared libraries.
configure: threading is disabled.
configure: internal memory pools for packing buffers are enabled.
configure: libmemkind not found; disabling.
configure: the BLAS compatibility layer is enabled.
configure: the CBLAS compatibility layer is disabled.
configure: the internal integer size is automatically determined.
configure: the BLAS/CBLAS interface integer size is 32-bit.
configure: configuring for conventional gemm implementation.
configure: creating ./config.mk from ./build/config.mk.in
configure: creating ./bli_config.h from ./build/bli_config.h.in
configure: creating ./obj/generic
configure: creating ./obj/generic/config/generic
configure: creating ./obj/generic/kernels/generic
configure: creating ./obj/generic/ref_kernels/generic
configure: creating ./obj/generic/frame
configure: creating ./obj/generic/blastest
configure: creating ./obj/generic/testsuite
configure: creating ./lib/generic
configure: creating ./include/generic
configure: mirroring ./config/generic to ./obj/generic/config/generic
configure: mirroring ./kernels/generic to ./obj/generic/kernels/generic
configure: mirroring ./ref_kernels to ./obj/generic/ref_kernels
configure: mirroring ./ref_kernels to ./obj/generic/ref_kernels/generic
configure: mirroring ./frame to ./obj/generic/frame
configure: creating makefile fragments in ./obj/generic/config/generic
configure: creating makefile fragments in ./obj/generic/kernels/generic
configure: creating makefile fragments in ./obj/generic/ref_kernels
configure: creating makefile fragments in ./obj/generic/frame
configure: configured to build within top-level directory of source distribution.
[jhammond@vestalac2 blis]$ make
Generating monolithic blis.h.................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................Traceback (most recent call last):
  File "./build/flatten-headers.py", line 515, in <module>
    main()
  File "./build/flatten-headers.py", line 496, in main
    final_string = strip_cstyle_comments( final_string )
  File "./build/flatten-headers.py", line 201, in strip_cstyle_comments
    return re.sub( "/\*.*?\*/", "", string, flags=re.S )
TypeError: sub() got an unexpected keyword argument 'flags'
make: *** [include/generic/blis.h] Error 1
```

**Comment by jeffhammond** [comment]
If you have a Python version dependency, you need to implement a check for it.
```sh
[jhammond@vestalac2 blis]$ python 
Python 2.6.6 (r266:84292, Aug  9 2016, 06:13:05) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-17)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
```

**Comment by jeffhammond** [comment]
Yep, BLIS assumes Python 2.7.  Once I use it, BLIS compiles fine.

**Comment by rvdg** [comment]
I am negotiating with IBM to provide kernels for the Power architectures.  However, management there seems to want to support ATLAS instead...

**Comment by fgvanzee** [comment]
@jeffhammond Re python: I agree. I'll add it to the pile. (python 2.7+ is required.)
Re gcc: BLIS currently requires gcc 4.7+.
Re clang: see python comment.

**Comment by cdluminate** [comment]
I have root access to a ppc64el machine and I'll try to debug.

**Comment by cdluminate** [comment]
@rvdg 
https://github.com/opencv/opencv/issues/7207#issuecomment-249863582

**Comment by rvdg** [comment]
@cdluminate Thanks for the pointer.  I prefer working my contacts, in this case the person who for decades has written the dgemm implementation for benchmarking large IBM systems.  His implementations, surprisingly, often then did not make it into IBM's ESSL library, even though they were faster.  There is weird politics at work.  Past experience showed that beating IBM's ESSL library is easy with a library like BLIS.

**Comment by loveshack** [comment]
In case you don't know, and it helps, IBM seem to have sponsored OpenBLAS work.
I guess #259 is relevant in the meantime.

**Comment by rvdg** [comment]
I am in communication with "people"  (good people) at IBM.  They have requested permission to contribute to BLIS.  This is working its way through IBM's bureaucracy.  For a few months now...  It is worth waiting for this to play out, since it would ensure very good microkernel.  It appears that IBM's strategy so far has been to support OpenBLAS and ATLAS.

**Comment by loveshack** [comment]
Good!  I only mentioned OpenBLAS because it sounded as if one hand was
just supporting ATLAS; apologies for the noise.

**Comment by cdluminate** [comment]
FYI the `generic` configuration works fine on ppc64el

https://buildd.debian.org/status/fetch.php?pkg=blis&arch=ppc64el&ver=0.5.1-1&stamp=1547188360&raw=0

**Comment by loveshack** [comment]
> FYI the `generic` configuration works fine on ppc64el

Sure, that's what the Fedora packaging uses, but it probably runs at a
similar speed to reference BLAS; I raised that in another issue.

By the way, is there a reason the Debian packaging doesn't provide
libblas with alternatives, as Debian openblas and Fedora blis do?
(Presumably that's only worthwhile for x86_64, and possibly aarch64.)

**Comment by cdluminate** [comment]
> By the way, is there a reason the Debian packaging doesn't provide libblas with alternatives, as Debian openblas and Fedora blis do? (Presumably that's only worthwhile for x86_64, and possibly aarch64.)

This is a false proposition. AFAIK the latest packaging for Debian's BLIS is quite complete and there are 6 flavors to install. And most the update-alternative mechanism is in fact available for 20+ architectures as long as the package compiles. https://tracker.debian.org/pkg/blis

6 flavours = (32-bit index, 64-bit index) x (omp, pthread, serial)

Additionally, libblis2-* provides libblas.so.3, and libblis64-2-* provides libblas64.so.3. In contrary, OpenBLAS's packaging only provide the (32-bit index + pthread) flavour. I've already coordinated with OpenBLAS maintainer to apply the packaging layout of BLIS to OpenBLAS in the future.

The 64-bit index BLAS libraryes are still WIP and experimental. See bottom part of
https://wiki.debian.org/DebianScience/LinearAlgebraLibraries

**Comment by loveshack** [comment]
You wrote: 

> This is a false proposition. AFAIK the latest packaging for Debian's
> BLIS is quite complete and there are 6 flavors to install. And most
> the update-alternative mechanism is in fact available for 20+
> architectures as long as the package
> compiles. https://tracker.debian.org/pkg/blis
>
> 6 flavours = (32-bit index, 64-bit index) x (omp, pthread, serial)

I didn't mean that.  I saw the combinatorial builds, which is good
compared with the current state with openblas.  However, I now realize
they're not installable together, which is unfortunate.

I'd hoped to coordinate packaging for the distributions to make things
consistent as much as possible.

> Additionally, libblis2-* provides libblas.so.3, and libblis64-2-*
> provides libblas64.so.3. 

Good, but I couldn't see them -- either in the source or the package
contents; that's what I was commenting on.  Anyhow, I guess this is
off-topic here unless blis people are interested in what distributions
are providing, or should provide.

**Comment by loveshack** [comment]
I'll follow up on the Debian tracker about the packaging.  However, it's
perhaps worth making the point here that the different threading options
need to be installable in parallel, and the default should be serial --
despite MKL.  That's from the point of view of both a distribution
packager and research computing person.

**Comment by loveshack** [comment]
It occurs to me, that if IBM have only contributed POWER9 support and
there's a decent chance the POWER7 support will work, I could at least
test it on POWER8.  I couldn't fix anything non-trivial at the assembler
level, though.

**Comment by cdluminate** [comment]
> However, it's perhaps worth making the point here that the different threading options need to be installable in parallel, and the default should be serial -- despite MKL. That's from the point of view of both a distribution packager and research computing person.

But I still don't understand why different threading versions should be co-installable. As Debian doesn't change the library name like fedora, installing two different versions is very likely to confuse normal users because they cannot tell which on earth is really used, especially when they don't know how to deal with the complicated 2-level alternatives. (there will be more than three layers of symbol links). Besides, Rrenaming libblis into libbliso, libblisp, libbliss is breaking the advantage of alternatives, and looks messy.

Disallowing co-installation reduces my maintainance burden due to reduced alternative level. And I can assume that people who use BLIS intentionally has at least foundamental background on BLAS/LAPACK.

**Comment by fgvanzee** [comment]
> However, it's perhaps worth making the point here that the different threading options need to be installable in parallel

Perhaps what @loveshack is suggesting is a configuration option to BLIS that activates both the OpenMP and pthreads components simultaneously, with some default (OpenMP maybe?) and the ability for the user to change the threading model that is employed at runtime? (e.g. `bli_thread_set_model()`.)

I'm not exactly itching to implement this dual-threading strategy, but I'm willing to entertain that it might be useful to some people.

**Comment by fgvanzee** [comment]
Sorry for the confusion, then. I'll leave this discussion to the package people, then. :)

**Comment by devinamatthews** [comment]
The discussion seems to have devolved somewhat from the original issue. We had an undergraduate working on Power8 kernels with some guidance from on high but I don't think it's officially merged yet. Closing until that gets mainlined and any other issues pop up.
